### PR TITLE
Add support for currency fields in CSV files

### PR DESCRIPTION
**Pull request by kubauk**
I would like to introduce functionality that will parse and understand currency fields in CSV files. Many of us have transactions in multiple currencies and there are more and more financial institutions that provide multi-currency accounts.

The most contentious change is in the last comment where I modify a condition from __balance is not None__ to __not balance__. This change means where before empty balance fields would produce a zero balance assertion, they will now produce none. 

This test seems more correct to me as it also test for empty strings as well as None. There are many other tests for other variables and fields that are still is not None which I think may have meant to be not <variable> but I have not addressed those in this PR.

### CODE REVIEW

**Comment by dnicolodi** [comment]
The change regarding the balance directives seems wrong to me. The `is not None` check is used to distinguish between a value not set and a numeric value that may be zero thus false in boolean context. I haven't looked carefully at the code, but at that stage, `balance` should be a `Decimal` object, thus I don't understand the comment about the empty string.

Can you also please squash the commits into a single commit?

File: beangulp/importers/csv.py [hunk 1]
```diff
@@ -274,6 +277,9 @@ def get(row, ftype):
                 last4 = get(row, Col.LAST4)
 
                 balance = get(row, Col.BALANCE)
+                currency = self.currency \
+                    if Col.CURRENCY not in iconfig or not get(row, Col.CURRENCY) \
+                    else get(row, Col.CURRENCY)
```
Comment (dnicolodi) [inline beangulp/importers/csv.py:282]: It seems that this would be easier to unwrap into a regular `if ... else` block.

File: beangulp/importers/csv.py [hunk 2]
```diff
@@ -73,6 +73,9 @@ class Col(enum.Enum):
     # account.
     CATEGORY = '[CATEGORY]'
 
+    # A column that indicates the amount currency for the current row which may be different to the base currency.
```
Comment (dnicolodi) [inline beangulp/importers/csv.py:76]: This line is to long. Can you please wrap the comment a 78 columns or there about?

File: beangulp/importers/csv.py [hunk 3]
```diff
@@ -327,14 +333,23 @@ def get(row, ftype):
...
+            entry = None
+            if Col.CURRENCY in iconfig:
+                for e in entries[::-1]:
+                    balance = e.meta.get('balance', None)
+                    if balance and balance[1] == self.currency:
+                        entry = e
+                        break
+
+            if not entry:
+                entry = entries[-1]
```
Comment (dnicolodi) [inline beangulp/importers/csv.py:345]: I have the impression that this does not do what intended if there are no transactions in the main currency in the CSV file. Also, wouldn't it be better to emit one `balance` directive for each currency in the file?

Comment (kubauk) [reply beangulp/importers/csv.py:345]: Yes I was also thinking about producing a balance per currency however I stopped at one because in my specific case I did not need it to do so. If you think it's required I can put the functionality in.

**Comment by kubauk** [comment]
The balance change in an interesting one. For some reason, I had decided that sometimes CSV could have a balance field but the contents could be empty. In this case the change I added is needed as both None and '' are indications of no balance being present.

I'm now thinking I should revert the change and not cover this scenario as it really should not happen. Why would an institution put the field into the CSV without then populating it with valid data?

**Comment by kubauk** [comment]
I've ended up adding support for multiple balance assertions and I'm actually much happier with the code now. It's far more readable and just seems like the right thing to do in this case. 

I've also reverted the change around the balance test on the basis of my last comment. I just don't see the case it was catering for to be a valid one.

File: beangulp/importers/csv.py [hunk 4]
```diff
@@ -283,7 +292,7 @@ def get(row, ftype):
                 if txn_time is not None:
                     meta['time'] = str(dateutil.parser.parse(txn_time).time())
                 if balance is not None:
-                    meta['balance'] = self.parse_amount(balance)
+                    meta['balance'] = (self.parse_amount(balance), currency)
```
Comment (dnicolodi) [inline beangulp/importers/csv.py:295]: This would look more idiomatic if written
```python
meta['balance'] = Amount(self.parse_amount(balance), currency)
```
and the rest of the code adjusted to this change.

File: beangulp/importers/csv.py [hunk 5]
```diff
@@ -325,16 +334,22 @@ def get(row, ftype):
         if not is_ascending:
             entries = list(reversed(entries))
 
+        balances = {}
         # Add a balance entry if possible
-        if Col.BALANCE in iconfig and entries:
-            entry = entries[-1]
+        for entry in entries[::-1]:
```
Comment (dnicolodi) [inline beangulp/importers/csv.py:339]: I think the `if Col.BALANCE in iconfig` part should stay, to avoid iterating the entries if it is known that there are no balances.

Comment (kubauk) [reply beangulp/importers/csv.py]: That make sense. I will change it back to include the test.

File: beangulp/importers/csv.py (hunk 4 above)
Comment (kubauk) [reply beangulp/importers/csv.py:295]: I missed this in the last push. I will add the changes now.

**Comment by dnicolodi** [comment]
I now committed a slightly cleaned up version of this patch. Please test and let me know.
